### PR TITLE
Feature blog for KEP-3017: Unhealthy Pod Eviction Policy for PDBs

### PR DESCRIPTION
**Pull request by atiratree**
enhancement: https://github.com/kubernetes/enhancements/issues/3017

API changes + feature gate: https://github.com/kubernetes/kubernetes/pull/113375

### CODE REVIEW

**Comment by fsmunoz** [comment]
Hello @atiratree , v1.26 Comms lead here. This feature blog is on a feature tracked for release, the deadline for submitting a draft is the **29th of November**; this should be considered the hard limit since we will need to review/edit/discuss the draft, so if at all possible it's better to submit it earlier to avoid any problems.

Any doubts, we're here to help!

**Review by Debanitrkl — COMMENTED** [review]
Hi @atiratree, v1.26 release comms shadow here, as the deadline is already over and the feature blog PR is not up yet, should be consider it cancelled for this release.

**Comment by sftim** [comment]
You can still blog about the new feature, just not as part of the post-release blog series for v1.26.

Please do consider tweaking the article for publication early next year.

File: content/en/blog/_posts/2022-12-08-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [inline content/en/blog/_posts/2022-12-08-unhealthy-pod-eviction-policy-for-pdb.md:68]: Not sure what to do here. The KEP for this feature is not up to date. I am trying to address that with https://github.com/kubernetes/enhancements/pull/3677 , but not sure how fast it will merge.

The enhancements link here is for the new KEP and is not working currently.

**Comment by atiratree** [comment]
@Debanitrkl @fsmunoz @sftim sorry for the constant delays. I have pushed the proposal and will try to find people to review this. 

Not sure if it can still make it to the post-release blog series, but I will be happy to see this published even on a later date.

**Comment by fsmunoz** [comment]
Thank you @atiratree , we're looking at it!

File: content/en/blog/_posts/2022-12-08-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2022-12-08-unhealthy-pod-eviction-policy-for-pdb.md]: we have decided not to rename the KEP link in https://github.com/kubernetes/enhancements/pull/3677 to not to cause any disruptions

**Comment by fsmunoz** [comment]
@atiratree publication date for this feature blog is **January 6, 2023**, you can update the date in the header. We have the series in good shape so I don't see a problem in adding this one, regardless of the missed deadlines.

The publication date is still relatively far away, but we should be able to review and merge it before, assuming that comments and suggestions are quickly reflected - your help in that is greatly appreciated.

File: content/en/blog/_posts/2022-12-08-unhealthy-pod-eviction-policy-for-pdb.md
Comment (onlydole) [inline content/en/blog/_posts/2022-12-08-unhealthy-pod-eviction-policy-for-pdb.md:4]:
```suggestion
date: 2022-01-06
```

Comment (onlydole) [reply content/en/blog/_posts/2022-12-08-unhealthy-pod-eviction-policy-for-pdb.md:4]: Suggesting a new date for when the 1.26 blog series kicks off.

Comment (sftim) [reply content/en/blog/_posts/2022-12-08-unhealthy-pod-eviction-policy-for-pdb.md:4]: Please also rename `content/en/blog/_posts/2022-12-08-unhealthy-pod-eviction-policy-for-pdb.md` to match.

**Comment by atiratree** [comment]
Thank you. I have updated the date

**Review by sftim — COMMENTED** [review]
Hi. Here's some more feedback.

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (soltysh) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:9]: @atiratree maybe something like:

Ensuring the disruptions to your applications do not affect its availability isn't a simple task. With 
the introduction of Unhealthy Pod Eviction Policy for Pod Disruption Budgets this is going to change. 
In this article, we will dive deeper into what modifications were introduced in PDBs to give application 
owners greater flexibility in managing disruptions.

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (soltysh) [inline content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:12]: Link to https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-budgets

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (sftim) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:9]:
```suggestion

Ensuring the disruptions to your applications do not affect its availability isn't a simple
task. With the introduction of _unhealthy pod eviction policy_ for PodDisruptionBudgets,
the way you can ensure this has changed again.
In this article, we will dive deeper into what modifications were introduced for PDBs to
give application owners greater flexibility in managing disruptions.

```
?

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (sftim) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:12]: If making the change suggested in https://github.com/kubernetes/website/pull/37627/files#r1053221368, you could put the link there instead.

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md]: Added suggestion. 

Somebody could manage that, but it is not advisable in general since only the application owner would know if that is okay.

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:9]: @soltysh thanks for the suggestion. I have added it there, with the link to PDBs.


>  With the introduction of _unhealthy pod eviction policy_ for PodDisruptionBudgets,
the way you can ensure this has changed again.

@sftim not sure what his is trying to say. Can we keep Maciej's version?


> Please also mention that this is describing a feature released last month, ie December 2022.


Do we need to mention the time, when we specify the version in the title?

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:51]: I am not sure if I follow.

we are describing the behaviour of the default IfHealthyBudget in the section above with.

```
Eviction of pods respects PodDisruptionBudgets (PDBs). This means that the eviction of a pod
should not disrupt a guarded application and `.status.currentHealthy` of a PDB should not fall 
below `.status.desiredHealthy`. Running pods that are [Unhealthy](/docs/tasks/run-application/configure-pdb/#healthiness-of-a-pod)
do not count towards the PDB status, but eviction of these is only possible in case the application
is not disrupted. This helps disrupted or not yet started application to achieve availability
as soon as possible without additional downtime that would be caused by evictions.
```

or are you suggesting we should describe the edge cases of Eviction API?

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md]: I have put the `Kubernetes 1.26` just for now. I like the rest ` Configure How PodDisruptionBudgets Affect Unhealthy Pods` , but I am wondering if we could/should mention the eviction part as well?

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md]: I guess, it is a good idea to suggest this. I have made small modifications to your text.

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md]: +1 for describing it more. But I wouldn't go as far as to suggest `you know that this app doesn't manage local
state` since this appies only to a subset of such applications.

I have put `You want to limit avoidable disruption, and you know that best effort availability is sufficient for this app.` there instead.

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md]: added the link to the intro

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (sftim) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:9]: Not sure what you mean by time? We don't need to be precise to the hour.

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (sftim) [inline content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:13]: ”going to change” is what puzzles me. 1.26 is already out.

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (sftim) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:51]: If:
- you have a workload with _n_ Pods
- you protect the workload with a PodDisruptionBudget
- you choose, or default to, `IfHealthyBudget`
- several of those Pods are in CrashLoopBackoff

and you then
- (attempt to) drain a node where the CrashLoopBackoffs are happening

what actually happens? It'd be good to summarise this mainly in terms of `kubectl drain`, but readers might also want to know at API level. For example: you try to create an eviction, and that request fails with error 418, `I'm a teapot` (OK, probably not that specific error code).

My ask here is to explain the outcome that you _don't_ see if you enable the feature gate and set a custom `unhealthyPodEvictionPolicy`, so that readers can appreciate the benefit.

**Comment by sftim** [comment]
We can merge this if needed; I'd prefer some more polish first.

**Comment by fsmunoz** [comment]
@atiratree if you could take a look at the added review comments, it would be great. Publication date is 4 days ahead, so there's still time. Otherwise, we can default to merge as-is, but this would be a last option.

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:51]: I see, I have tried to put more details to describe the behavior.

```
Unfortunately, this poses a problem for cluster administrators that would like to drain nodes
without any manual interventions. Misbehaving applications with pods in `CrashLoopBackOff`
state (due to a bug or misconfiguration) or pods that are simply failing to become ready
make this task much harder. Any eviction request will fail due to violation of a PDB, 
when all pods of an application are unhealthy. Draining of a node cannot make any progress
in that case.
```

Do you think it make sense to describe the exact message and http code here?

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:9]: sorry, I meant the date and not the time. I have rewritten the `going to change` part. Please let me know if you have something better in mind.

Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:9]: and we also mention the k8s version in the `What problems does this solve?` section

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:13]: https://github.com/kubernetes/website/pull/37627#discussion_r1060182380

**Comment by atiratree** [comment]
@sftim I am trying to address the missing pieces. can you please take a look again?

**Comment by tengqm** [comment]
@atiratree Please go ahead and revise the text.

**Comment by atiratree** [comment]
@tengqm do you have anything in particular in mind? I believe I have replied to all of the remarks so far.

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (sftim) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:13]:
```suggestion
has changed.
```

That's a simple rewrite; you can use other words, so long as we avoid the incorrect use of future tense.

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (sftim) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:51]: At this point, you should describe _something_. I haven't tried the feature out so I don't know what you need to say. I don't think “Eviction API will consider these policies when eviction of a pod that is guarded by a PDB is requested.” feels like enough.

That said, if we had to we could merge this and live with the confusion.

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (sftim) [inline content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:13]: Another option:
```suggestion
task. Last month's release of Kubernetes v1.26 lets you specify an  _unhealthy pod eviction policy_
for [PodDisruptionBudgets](/docs/concepts/workloads/pods/disruptions/#pod-disruption-budgets) (PDBs)
to help you maintain that availability during node management operations.

```

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (sftim) [inline content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:19]: (nit / optional change)
```suggestion
Eviction of pods respects PodDisruptionBudgets (PDBs). This means that a
[voluntary disruption](https://kubernetes.io/docs/concepts/scheduling-eviction/#pod-disruption)
to a Pod
```

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md]: +1, changed to this

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md]: Thx for the suggestion, I used part of it and reorganized the section: https://github.com/kubernetes/website/compare/4629fb1dcda1ab70deaed7a30f89228d0d1fbb6d..e27c6086386bcbdc67b52cf7c9909d3c5caf788d.

I think we should not mention the node-pressure eviction here, since it could distract from the main topic. The node-pressure eviction does not abide by the PDBs. I have tried to highlight that we are dealing with the API initiated eviction.

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md]: using the other option now: https://github.com/kubernetes/website/pull/37627#discussion_r1060794688

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (atiratree) [reply content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md]: I suppose we could link that. What do you think about still having the eviction mentioned there?

```
This means that a requested [voluntary disruption](https://kubernetes.io/docs/concepts/scheduling-eviction/#pod-disruption)
via an eviction to a Pod, should not disrupt a guarded application and `.status.currentHealthy` of a PDB should not fall
below `.status.desiredHealthy`
```

**Comment by atiratree** [comment]
@sftim @soltysh thanks for the feedback. Can you please do another round?

File: content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md
Comment (soltysh) [inline content/en/blog/_posts/2023-01-06-unhealthy-pod-eviction-policy-for-pdb.md:53]: Even though this does match the current behavior, I'd add a one sentence how it works. Something like:

Unhealthy pods can be disrupted only if the applications has a minimum available `.status.desiredHealthy` number of pods.

**Review by soltysh — APPROVED** [review]
two more nits, but otherwise this looks great to me

**Review by sftim — COMMENTED** [review]
Let's merge it. @atiratree even after merge, you can open a PR to make minor edits - eg, see https://github.com/kubernetes/website/pull/37627#pullrequestreview-1237117336

/lgtm
/approve

**Comment by atiratree** [comment]
@soltysh @sftim good idea, I am adressing the nits in https://github.com/kubernetes/website/pull/38784

**Pull request merged**
